### PR TITLE
Fix video recorder hanging after fixture skip

### PR DESCRIPTION
**Pull request by kisrefod**
## Purpose
_When local testcafe use video-recorder ffmpeg and skip any fixture - it doesn't terminate in Mac and Linux_

## Approach
_In this commit changed kill algorithm for unix-processes, added call "test-before-done", that finish ffmpeg capturing and removed ffmpeg call whenfixture skipped. Tests for that issue already in this fix_

## References
_https://github.com/DevExpress/testcafe/issues/6163_

## Pre-Merge TODO
- [ ] Check changes
- [ ] Check tests for that changes

### CODE REVIEW

File: test/server/kill-test.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
+const __importDefault = this && this.__importDefault || function (mod) {
+    return mod && mod.__esModule ? mod : { 'default': mod };
+};
+const sinon     = require('sinon');
+const assert    = require('assert');
+const osFamily = __importDefault(require('os-family'));
```
Comment (AndreyBelym) [inline test/server/kill-test.js:6]: `__importDefault` is an auto-generated construct produced by compiler. Just use `require`.

File: src/utils/process.js [hunk 2]
```diff
@@ -65,30 +64,41 @@ async function findProcessUnix (browserId) {
...
+        await delay(CHECK_KILLED_DELAY);
+        unixProcessKilled = await isUnixProcessKilled(processId);
+    } while (!unixProcessKilled && softTries < 2);
 
-    const killTimeoutTimer = delay(CHECK_PROCESS_IS_KILLED_TIMEOUT)
-        .then(() => {
-            timeoutError = true;
-        });
+    unixProcessKilled = await isUnixProcessKilled(processId);
+    if (unixProcessKilled) return;
```
Comment (AndreyBelym) [inline src/utils/process.js:93]: Please, add more whitespace to make your code consistent with the all other code and improve readability:
```suggestion
    let softTries         = 0;
    let unixProcessKilled = false;

    do {
        await killUnixProcessSoft(processId);
        
        softTries++;
        
        await delay(CHECK_KILLED_DELAY);
        
        unixProcessKilled = await isUnixProcessKilled(processId);
    } 
    while (!unixProcessKilled && softTries < 2);

    unixProcessKilled = await isUnixProcessKilled(processId);
    
    if (unixProcessKilled)
        return;
```

File: src/video-recorder/process.js [hunk 3]
```diff
@@ -143,23 +150,25 @@ export default class VideoRecorder extends AsyncEmitter {
 
         await delay(FFMPEG_START_DELAY);
     }
+    async dispose () {
```
Comment (AndreyBelym) [inline src/video-recorder/process.js:153]: We separate different methods and functions with an empty line:
```suggestion
    }

    async dispose () {
```

File: test/server/kill-test.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                    if (event === 'data' && getInfoCount === 0) {
+                        listener('1 1');
+                        getInfoCount++;
+                    }
+                };
+
+                let unixProcessKilled = false;
+
+                const stubSpawn = sinon.stub(childProcess, 'spawn').returns(stubChildProcess);
+                const stubKill = sinon.stub(process, 'kill').callsFake((processId, flag) => {
```
Comment (AndreyBelym) [inline test/server/kill-test.js:63]: We separate multiline assignments from the singleline ones:
```suggestion
                const stubSpawn = sinon.stub(childProcess, 'spawn').returns(stubChildProcess);
                
                const stubKill = sinon.stub(process, 'kill').callsFake((processId, flag) => {
```

File: src/utils/process.js [hunk 5]
```diff
@@ -65,30 +64,41 @@ async function findProcessUnix (browserId) {
...
+    let softTries = 0;
+    let unixProcessKilled = false;
 
-    process.kill(processId);
+    do {
+        await killUnixProcessSoft(processId);
+        softTries++;
+        await delay(CHECK_KILLED_DELAY);
+        unixProcessKilled = await isUnixProcessKilled(processId);
+    } while (!unixProcessKilled && softTries < 2);
```
Comment (AndreyBelym) [inline src/utils/process.js:90]: Please, use constants for string and number literals, except -1,0,1 and ''.

So could you please extract 'SIGKILL' and 2 to constants?

**Review by AndreyBelym — COMMENTED** [review]
It's good in general, but we should work on codestyle.

File: src/utils/process.js [hunk 6]
```diff
@@ -65,30 +65,49 @@ async function findProcessUnix (browserId) {
...
-            throw new Error(killProcessTimeoutError);
-    });
+        await delay(CHECK_KILLED_DELAY);
+
+        unixProcessKilled = await isUnixProcessKilled(processId);
+    } while (!unixProcessKilled && softTries < maxSoftTries);
+
+    unixProcessKilled = await isUnixProcessKilled(processId);
+
+    if (unixProcessKilled) return;
```
Comment (AndreyBelym) [inline src/utils/process.js:99]: Please always place blocks in complex operators on separate lines
```suggestion
    if (unixProcessKilled) 
        return;
```

File: src/utils/process.js [hunk 7]
```diff
@@ -65,30 +65,49 @@ async function findProcessUnix (browserId) {
...
+
+    await killUnixProcessHard(processId);
+
+    await delay(CHECK_KILLED_DELAY);
+
+    unixProcessKilled = await isUnixProcessKilled(processId);
+
+    if (unixProcessKilled) return;
+
+    //if 2 soft-kill and 1 hard-kill with "SIGKILL"-flag didn't work - throw error
```
Comment (AndreyBelym) [inline src/utils/process.js:109]: Please prefix your commits with `NOTE:`, `HACK:` or `TODO: `.

File: src/video-recorder/recorder.js [hunk 8]
```diff
@@ -132,13 +132,18 @@ export default class VideoRecorder extends EventEmitter {
...
             this.warningLog.addWarning(WARNING_MESSAGES.videoNotSupportedByBrowser, testRunVideoRecorder.testRunInfo.alias);
+            return;
+        }
+
+        const isVideoEnabled = await testRunVideoRecorder.isVideoEnabled();
+
+        if (!isVideoEnabled)
+            return;
+
+        await testRunVideoRecorder.init();
```
Comment (AndreyBelym) [inline src/video-recorder/recorder.js:145]: It's better to separate `await` statements from synchronous code with empty lines

File: src/utils/process.js [hunk 9]
```diff
@@ -65,30 +65,49 @@ async function findProcessUnix (browserId) {
...
+        softTries++;
 
-    return Promise.race([killTimeoutTimer, checkUnixProcessIsKilled(processId)]).then(() => {
-        if (timeoutError)
-            throw new Error(killProcessTimeoutError);
-    });
+        await delay(CHECK_KILLED_DELAY);
+
+        unixProcessKilled = await isUnixProcessKilled(processId);
+    } while (!unixProcessKilled && softTries < maxSoftTries);
```
Comment (AndreyBelym) [inline src/utils/process.js:95]: We use the `stroustoup` style for code blocks: https://eslint.org/docs/rules/brace-style#stroustrup

```suggestion
    do {
        await killUnixProcessSoft(processId);

        softTries++;

        await delay(CHECK_KILLED_DELAY);

        unixProcessKilled = await isUnixProcessKilled(processId);
    } 
    while (!unixProcessKilled && softTries < maxSoftTries)
```

File: src/video-recorder/test-run-video-recorder.js [hunk 10]
```diff
@@ -61,7 +61,9 @@ export default class TestRunVideoRecorder {
 
         return connectionCapabilities && connectionCapabilities.hasGetVideoFrameData;
     }
-
+    async isVideoEnabled () {
```
Comment (AndreyBelym) [inline src/video-recorder/test-run-video-recorder.js:64]: Please try to separate class and object methods with an empty line here and in other places.

**Review by AndreyBelym — COMMENTED** [review]
Some more code style, please 👍

**Review by AndreyBelym — APPROVED** [review]
Awesome work

**Review by AlexKamaev — APPROVED** [review]

**Pull request merged**